### PR TITLE
Loan disbursement added as savings transaction

### PR DESCRIPTION
**Pull request by nikIntrasoft**
Description of the issue/feature this PR addresses:

**Current behavior before PR:**
Currently on loan disbursement to savings account , the transaction type is shown as 'deposit'.

**Desired behavior after PR is merged:**
After merging this PR, the transaction of disbursing the loan to savings account would be shown as 'loan disbursement'.

--

### CODE REVIEW

**Comment by ottoman91** [comment]
The changes are fine from my end. @tarzan0820 we are good to merge .

**Comment by nikIntrasoft** [comment]
Resolved conflicts

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/data/SavingsAccountTransactionEnumData.java [hunk 1]
```diff
@@ -91,6 +95,10 @@ public boolean isDeposit() {
     public boolean isWithdrawal() {
         return this.withdrawal;
     }
+    
+    public boolean isLoanDisbursement() {
+        return this.loanDisbursement;
+    }
 
     public boolean isDepositOrWithdrawal() {
         return this.deposit || this.withdrawal;
```
Comment (tarzan0820) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/data/SavingsAccountTransactionEnumData.java]: SavingsAccountTransactionEnumData.isDeposit is used important places example org/apache/fineract/accounting/journalentry/service/CashBasedAccountingProcessorForSavings.java which contains logic that will breakdown with loan disbursement paid to savings. 

I think a better implementation will be:

this.deposit = Long.valueOf(SavingsAccountTransactionType.DEPOSIT.getValue()).equals(this.id) || Long.valueOf(SavingsAccountTransactionType.LOAN_DISBURSEMENT.getValue()).equals(this.id);

so you do not have to worry about changing the logic of all affected areas

**Comment by tarzan0820** [comment]
@nikIntrasoft I am not comfortable with this implementation given that checks for SavingsAccountTransactionType.DEPOSIT is littered all over the code base which means we would have to hunt all these down and introduce and introduce SavingsAccountTransactionType.LOAN_DISBURSEMENT. Also it is not future proof as changes made by upstream related to SavingsAccountTransactionType.DEPOSIT would break our implementation.

What we want to do is have an indication of what deposit is a loan disbursement and change the label from deposit to loan disbursement accordingly. 

An implementation can be as follows:

-  Introduce a boolean database column in the pertinent table for deposits that are disbursements from loan accounts. i think m_savings_account_transaction
- handleLoanDeposit can now set this flag after handleDeposit is called... i think this can be done by operating on the SavingsAccountTransaction object returned by  handleDeposit
- In org/apache/fineract/portfolio/savings/service/DepositAccountReadPlatformServiceImpl.java ln 997, SavingsAccountTransactionData mapRow, we can then check for this flag and if present send SavingsAccountTransactionType.LOAN_DISBURSEMENT as the transaction type to SavingsEnumerations.transactionType ... also note we might  (please check) have to effect this also for org/apache/fineract/portfolio/savings/service/DepositAccountReadPlatformServiceImpl.java ln 1445 and all other places that call this function (i see 4 others)

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/SavingsAccountTransactionType.java [hunk 2]
```diff
@@ -201,15 +211,21 @@ public boolean isAmountOnHold() {
...
+    
+    
     public boolean isDebit() {
         return isWithdrawal() || isWithdrawalFee() || isAnnualFee() || isPayCharge() || isIncomeFromInterest() || isWithHoldTax() || isEscheat() || isAmountOnHold();
     }
 
     public boolean isCredit() {
-        return isDeposit() || isInterestPosting() || isDividendPayout() || isAmountRelease();
+
+        return isDeposit() || isInterestPosting() || isDividendPayout() || isAmountRelease()||isLoanDisbursement();
```
Comment (nikIntrasoft) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/SavingsAccountTransactionType.java]: yes it is redundant

Comment (nikIntrasoft) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/SavingsAccountTransactionType.java]: yes this redundant.. but it is harmless. I have removed it in updated PR

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/data/SavingsAccountTransactionEnumData.java (hunk 1 above)
Comment (nikIntrasoft) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/data/SavingsAccountTransactionEnumData.java]: if you observe the reference wherein SavingsAccountTransactionEnumData is used, most of the time SavingsEnumerations.transactionType() is being used. Which would return apropriate transaction enum data and would not have caused problems. Anyways this has been updated

**Comment by nikIntrasoft** [comment]
This implementation was made keeping in mind, minimum point of changes which would ease the burden of maintenance.  Also the handleloandeposit was aligned to handleDeposit to absorb any future changes readily . Internally isDeposit is equivalent to isLoanDisbursement and hence any checks elsewhere in the code would have been passed. Anyways I have scrapped this PR and have sent a new PR

The approach you have suggested above will not serve our purpose accurately but I understood what you wanted to convey.(modifying DepositAccountReadPlatformServiceImpl is not going to work for us) Keeping in mind your approach I have sent a new PR, hope that would work.

Lastly, You would have to manually add a column to m_savings_account_transaction, adding this change to migration scripts would again leave us out of sync with the apache script numbers. 
here is the insert statement for the change:
ALTER TABLE `m_savings_account_transaction` ADD COLUMN `isLoanDisbursement` TINYINT(1) NULL DEFAULT '0' AFTER `is_manual`;

**Comment by nikIntrasoft** [comment]
Keeping it open for comment reviews

**Pull request closed without merging**